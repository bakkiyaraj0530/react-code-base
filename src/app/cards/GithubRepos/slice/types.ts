import { ReposList } from 'types/Repo';

/* --- STATE --- */
export interface GithubRepositiesState {
  repositoryname: string;
  loading: boolean;
  error?: RepoErrorType | null;
  repositories: ReposList[];
}
export enum RepoErrorType {
  RESPONSE_ERROR = 1,
  REPOSITORY_NOT_FOUND = 2,
  REPOSITORYNAME_EMPTY = 3,
  HAS_NO_REPO = 4,
  GITHUB_RATE_LIMIT = 5,
}

export interface Pagination {
  current: number;
  pageSize: number;
}
/* 
  If you want to use 'ContainerState' keyword everywhere in your feature folder, 
  instead of the 'HomePageState' keyword.
*/
export type ContainerState = GithubRepositiesState;
