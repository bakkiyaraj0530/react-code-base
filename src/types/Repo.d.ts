
export interface ReposList {
  forkCount: number;
  name: string;
  stargazerCount: number;
}

export interface RepoListData {
  data: {
    viewer: {
      repositories: {
        nodes: Array<ReposList>
      }
    }
  }
}

export interface RepoData {
  data: {
    viewer: {
      repository: ReposList
    }
  }
}
