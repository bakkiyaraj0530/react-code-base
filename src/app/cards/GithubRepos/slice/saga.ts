import { call, put, select, takeLatest, delay } from 'redux-saga/effects';
import { selectRepositoryname } from './selectors';
import { githubRepoFormActions as actions } from '.';
import { ReposList, RepoListData, RepoData } from 'types/Repo';
import { RepoErrorType } from './types';
import { fetchData } from './api';

/**
 * Github repos request/response handler
 */
export function* getAllRepos() {
  yield delay(500);
  try {
    const repos: RepoListData = yield call(fetchData, '');
    if (repos?.data?.viewer?.repositories?.nodes?.length > 0) {
      // Add the repos list into repositoties store
      yield put(actions.reposLoaded(repos?.data?.viewer?.repositories?.nodes));
    } else if(repos?.data?.viewer === undefined) {
      yield put(actions.repoError(RepoErrorType.REPOSITORYNAME_EMPTY));
     } else {
      yield put(actions.repoError(RepoErrorType.REPOSITORY_NOT_FOUND));
     }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.repoError(RepoErrorType.REPOSITORY_NOT_FOUND));
    } else if (err.message === 'Failed to fetch') {
      yield put(actions.repoError(RepoErrorType.GITHUB_RATE_LIMIT));
    } else if (err.message === 'Failed to fetch') {
      yield put(actions.repoError(RepoErrorType.GITHUB_RATE_LIMIT));
    } else {
      yield put(actions.repoError(RepoErrorType.RESPONSE_ERROR));
    }
  }
}

/**
 * Fetch Repos by name -  request/response handler
 */

export function* getRepoByName() {
  yield delay(500);
  // Select Repository Name from store
  const repositoryname: string = yield select(selectRepositoryname);
  if (repositoryname.length === 0) {
    yield put(actions.repoError(RepoErrorType.REPOSITORYNAME_EMPTY));
    return;
  }

  try {
    const repository: RepoData = yield call(fetchData, repositoryname);

    const repodata: ReposList = repository?.data?.viewer?.repository;
    if (repodata !== null) {
      yield put(actions.repoLoadedByName(repodata));
    } else {
      yield put(
        actions.errorOnLoadingByName(RepoErrorType.REPOSITORY_NOT_FOUND),
      );
    }
  } catch (err) {
    if (err.response?.status === 404) {
      yield put(actions.repoError(RepoErrorType.HAS_NO_REPO));
    } else if (err.message === 'Failed to fetch') {
      yield put(actions.repoError(RepoErrorType.GITHUB_RATE_LIMIT));
    } else {
      yield put(actions.repoError(RepoErrorType.RESPONSE_ERROR));
    }
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* githubRepoFormSaga() {
  // Watches for loadRepos actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  // It will be cancelled automatically on component unmount
  yield takeLatest(actions.loadRepos.type, getAllRepos);
  yield takeLatest(actions.loadRepoByName.type, getRepoByName);
}
