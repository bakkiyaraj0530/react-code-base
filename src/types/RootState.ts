import { GithubRepositiesState } from 'app/cards/GithubRepos/slice/types';
import { ThemeState } from 'styles/theme/slice/types';


export interface RootState {
  theme?: ThemeState;
  githubRepoForm?: GithubRepositiesState;
}
