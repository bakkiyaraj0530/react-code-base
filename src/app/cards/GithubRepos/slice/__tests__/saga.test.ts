import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '..';

import { githubRepoFormSaga, getAllRepos } from '../saga';
import { RepoErrorType } from '../types';

describe('getAllRepos Saga', () => {
  let respositoryname: any;
  let repos: any;
  let getReposIterator: ReturnType<typeof getAllRepos>;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getReposIterator = getAllRepos();
    const delayDescriptor = getReposIterator.next().value;
    expect(delayDescriptor).toMatchSnapshot();

    const selectDescriptor = getReposIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should return error if repository is empty', () => {
    respositoryname = '';
    const putDescriptor = getReposIterator.next(respositoryname).value;
    expect(putDescriptor).toEqual(
      put(slice.githubRepoFormActions.repoError(RepoErrorType.REPOSITORYNAME_EMPTY)),
    );

    const iteration = getReposIterator.next();
    expect(iteration.done).toBe(true);
  });

  it('should dispatch the reposLoaded action if it requests the data successfully', () => {
    respositoryname = 'test';
   
    repos = [{
      forkCount: 1,
      name: 'test',
      stargazerCount: 1
    }]
    
    const requestDescriptor = getReposIterator.next(respositoryname).value;
    expect(requestDescriptor).toMatchSnapshot();
  });

  it('should dispatch the user not found error', () => {
    respositoryname = 'test';
    const requestDescriptor = getReposIterator.next(respositoryname).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.throw({ response: { status: 404 } })
      .value;
    expect(putDescriptor).toEqual(
      put(slice.githubRepoFormActions.repoError(RepoErrorType.REPOSITORY_NOT_FOUND)),
    );
  });

  it('should dispatch the github rate limit error', () => {
    respositoryname = 'test';

    const requestDescriptor = getReposIterator.next(respositoryname).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.throw(new Error('Failed to fetch'))
      .value;
    expect(putDescriptor).toEqual(
      put(
        slice.githubRepoFormActions.repoError(RepoErrorType.GITHUB_RATE_LIMIT),
      ),
    );
  });

  it('should dispatch the response error', () => {
    respositoryname = 'test';

    const requestDescriptor = getReposIterator.next(respositoryname).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = getReposIterator.throw(new Error('some error')).value;
    expect(putDescriptor).toEqual(
      put(slice.githubRepoFormActions.repoError(RepoErrorType.RESPONSE_ERROR)),
    );
  });
});

describe('githubRepoFormSaga Saga', () => {
  const githubRepoFormIterator = githubRepoFormSaga();
  it('should start task to watch for loadRepos action', () => {
    const takeLatestDescriptor = githubRepoFormIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.githubRepoFormActions.loadRepos.type, getAllRepos),
    );
  });
});
