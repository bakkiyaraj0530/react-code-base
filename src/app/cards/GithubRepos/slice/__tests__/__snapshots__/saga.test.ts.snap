// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAllRepos Saga should dispatch the github rate limit error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the github rate limit error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the github rate limit error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": 3,
      "type": "githubRepoForm/repoError",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getAllRepos Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the reposLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the reposLoaded action if it requests the data successfully 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": 3,
      "type": "githubRepoForm/repoError",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getAllRepos Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the response error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": 3,
      "type": "githubRepoForm/repoError",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getAllRepos Saga should dispatch the user not found error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the user not found error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should dispatch the user not found error 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "payload": 3,
      "type": "githubRepoForm/repoError",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`getAllRepos Saga should return error if repository is empty 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      500,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getAllRepos Saga should return error if repository is empty 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
